pdns/cachecleaner.hh

```diff
@@ -167,15 +172,15 @@ uint64_t pruneMutexCollectionsVector(C& container, std::vector<T>& maps, uint64_
         ++i;
       }
 
-      if (toTrim && erased >= toTrim / maps_size)
+      if (toTrim > 0 && erased >= toTrimForThisShard) {
```

Comment:
I wonder if it is smart to actually stop scanning for expired entries after we have cleaned up enough entries from this shard to meet our target. If we are under pressure (cache is full), we will likely end up cleaning just enough entries to get just below the maximum cache size, but not less, which means that the cache will very likely get full again before the next scan, and so on and so forth.
Perhaps we should keep scanning for expired entries even after meeting our target?

Reply:
See my comment above

Reply:
> Pushed a fixed version, with more explanation on the strategy used. One point to consider: in the first loop, we might ignore `toTrim` and just hunt for expired entries, using the `toScan` bound. That means that under pressure, we prefer to evict more expired entries and will fall back to the second loop in less cases. It might increase imbalance, but that _should_ not be an issue: we are removing stale items only.

I don't think it will increase imbalance, as we will scan the exact same number of entries for expired ones in every shard, since we do not take into account how many entries were removed so far for `toScan`. And if we do not remove enough entries that way, the second loop will nuke entries based on what is remaining at that point, so the balance should still be correct.

> Another thing to ponder:
> 
> Even while "data" record sets of a zone are retrieved and placed in the back of the expiry queue, the infra (NS + glue) record sets for the zone are not, as they are not touched when just checking the known "data" record sets of a zone. So infra record sets tend to travel to the front of the expiry queue, even when their data records are very popular. Only an expired data record set will trigger a retrieval of the infra record set from the cache so that it will be placed in the back of the expiry queue. So when the cache is under pressure it might clean more infra records than you would think.
> 
> This might be a case of "this is just the way it is", but IMO this is another reason we want to consider a separate cache fo infra records, with a different eviction policy.

My feeling is that it is working as expected: these infra records move to the front of the queue because we do not need them often, and thus we prefer to keep other records in the cache instead.

Reply:
Yes, but the potential impact of evicting infra records is higher compared to data records, as a cache miss of a data record that also has missing infra records induces more effort compared to just a missing data record.

Reply:
> I don't think it will increase imbalance, as we will scan the exact same number of entries for expired ones in every shard, since we do not take into account how many entries were removed so far for `toScan`. And if we do not remove enough entries that way, the second loop will nuke entries based on what is remaining at that point, so the balance should still be correct.

Right, as we take into account the size if the shard and we scan a fraction of t, not a fixed number as we used to.

I'll push a commit to that effect.

Reply:
> Yes, but the potential impact of evicting infra records is higher compared to data records, as a cache miss of a data record that also has missing infra records induces more effort compared to just a missing data record.

That's certainly true in some cases, but less so in others, for example because we do not DNSSEC-validate infra records unless asked to by a direct query. I'm currently not convinced a complex logic to keep the useful infra records in the cache would be worth it, but I might be wrong.